### ISSUE TITLE
canaper: Categorical Analysis of Neo- And Paleo-Endemism in R

### ISSUE THREAD

**Issue by joelnitta** [ropensci/software-review#475]
**Date accepted:** 2022-09-14
Submitting Author: @joelnitta
Other Package Authors: (delete if none) Shawn Laffan (@shawnlaffan)
Repository:  https://github.com/joelnitta/canaper
Version submitted: 0.0.1
Submission type: Stats
Badge grade: silver
Editor:  @tdhock
Reviewers: @KlausVigo, @luismurao
Due date for @KlausVigo: 2021-11-24
Due date for @luismurao: 2021-12-21
Archive: TBD
Version accepted: TBD

---

-   Paste the full DESCRIPTION file inside a code block below:

```
Package: canaper
Title: Categorical Analysis of Neo- And Paleo-Endemism
Version: 0.0.1
Authors@R: 
    c(
    person(given = "Joel H.",
           family = "Nitta",
           role = c("aut", "cre"),
           email = "joelnitta@gmail.com",
           comment = c(ORCID = "0000-0003-4719-7472")),
    person(given = "Shawn W.",
           family = "Laffan",
           role = c("ctb", "dtc")),
    person(given = "Brent D.",
           family = "Mishler",
           role = c("ctb", "dtc")),
    person(given = "Wataru",
           family = "Iwasaki",
           role = c("ctb"),
           comment = c(ORCID = "0000-0002-9169-9245"))
           )
Description: Provides functions to conduct categorical analysis of neo- and paleo-endemism (CANAPE).
License: MIT + file LICENSE
Encoding: UTF-8
LazyData: true
Roxygen: list(markdown = TRUE, roclets = c ("collate", "namespace", "rd", "roxyglobals::global_roclet", "srr::srr_stats_roclet"))
RoxygenNote: 7.1.2
URL: https://github.com/joelnitta/canaper
BugReports: https://github.com/joelnitta/canaper/issues
Imports: 
    ape,
    assertr,
    assertthat,
    dplyr,
    future.apply,
    phyloregion,
    progressr,
    purrr,
    stats,
    tibble,
    vegan
Suggests: 
    rmarkdown,
    knitr,
    future,
    tictoc,
    patchwork,
    tidyverse,
    testthat (>= 3.0.0),
    roxyglobals (>= 0.2.1),
    stringr,
    magrittr,
    covr,
    picante
Config/testthat/edition: 3
Depends: 
    R (>= 3.5.0)
VignetteBuilder: knitr
Remotes: 
    anthonynorth/roxyglobals
```

## Pre-submission Inquiry

- [x] A pre-submission inquiry has been approved in [issue #469](469)

## General Information

-   Who is the target audience and what are scientific applications of this package?
     
    - Target audience is evolutionary biologists. Scientific application is the analysis of phylogenetic endemism. 

-   Paste your responses to our [*General Standard* **G1.1** here](https://ropenscilabs.github.io/statistical-software-review-book/standards.html#general-standards), describing whether your software is:

    - *The first implementation within* **R** *of an algorithm which has previously been implemented in other languages or contexts*

    - Please include hyperlinked references to all other relevant software: https://github.com/shawnlaffan/biodiverse

-   (If applicable) Does your package comply with our [guidance around *Ethics, Data Privacy and Human Subjects Research*](https://devguide.ropensci.org/policies.html#ethics-data-privacy-and-human-subjects-research)?
    
    - (Not applicable)

Note also there is a package website available: https://joelnitta.github.io/canaper/

## Badging

- What grade of badge are you aiming for? ([bronze, silver, gold](https://ropenscilabs.github.io/statistical-software-review-book/pkgdev.html#pkgdev-badges)) 
    
    - Silver, but open to suggestions from the reviewers and editors.

- If aiming for silver or gold, describe which of the [four aspects listed in the *Guide for Authors* chapter](https://ropenscilabs.github.io/statistical-software-review-book/pkgdev.html#pkgdev-silver) the package fulfils (at least one aspect for silver; three for gold) 

    - Demonstrates excellence in compliance with multiple standards from at least two broad sub-categories
    - Compliance with a good number of standards beyond those identified as minimally necessary. I am in compliance with most of the standards for "General" and "Dimensionality Reduction, Clustering, and Unsupervised Learning". 

## Technical checks

Confirm each of the following by checking the box.

- [x] I/we have read the [guide for authors](https://devguide.ropensci.org/guide-for-authors.html) and [rOpenSci packaging guide](https://devguide.ropensci.org/building.html).
- [x] I/we have read the [*Statistical Software Peer Review* Guide for Authors](https://ropenscilabs.github.io/statistical-software-review-book/pkgdev.html).
- [ ] I/we have run [`autotest`](https://github.com/ropenscilabs/autotest) checks on the package, and ensured no tests fail.

     - `autotest` hangs (doesn't produce any output after waiting ca. 20 minutes) on most functions. I was only able to use it on `cpr_classify_endem()`.

- [x] The [`srr_stats_pre_submit()` function](https://ropenscilabs.github.io/srr/reference/srr_stats_pre_submit.html) confirms this package may be submitted.

This package:

- [x] does not violate the Terms of Service of any service it interacts with.
- [x] has a CRAN and OSI accepted license.
- [x] contains a [README with instructions for installing the development version](https://ropensci.github.io/dev_guide/building.html#readme).

## Publication options

- [x] Do you intend for this package to go on CRAN?
- [ ] Do you intend for this package to go on Bioconductor?

    - Also, I intend to submit a paper about this package to *Methods in Ecology and Evolution* after ROpenSci review.

## Code of conduct

- [x] I agree to abide by [rOpenSci's Code of Conduct](https://ropensci.github.io/dev_guide/policies.html#code-of-conduct) during the review process and in maintaining my package should it be accepted.

**Comment by mpadge** [comment]
@tdhock Only just realised the command there has the wrong syntax - should be `add <@name> to reviewers`, like [this example](https://github.com/ropensci/software-review/issues/470#issuecomment-954890209). Could you please try again? Nevertheless good that you got this one wrong, because it made us realise we need the bot to tell you it didn't understand. Thanks!

**Comment by mpadge** [comment]
sorry @tdhock, another bot glitch there - bot commands have to be kept as just that, with no additional comments. And yes, the point of the above-linked issue is to aid consistency between analogous commands, so that will hopefully improve soon. In the meantime, could you please try again, again, with just the command? With due apologies and gratitude! :bowing_man:

**Comment by tdhock** [comment]
@ropensci-review-bot add @KlausVigo to reviewers

**Comment by maelle** [comment]
@tdhock sorry for all the hiccups! The initial issue body had been edited after your username was added so your username was no longer in the string `<!--editor--> @tdhock<!--end-editor-->` I edited it again so you can try again. Thanks for your patience!

**Comment by tdhock** [comment]
@ropensci-review-bot add @KlausVigo to reviewers

**Comment by tdhock** [comment]
@ropensci-review-bot add @luismurao to reviewers

**Comment by luismurao** [comment]
## Package Review

*Please check off boxes as applicable, and elaborate in comments below.  Your review is not limited to these topics, as described in the reviewer guide*

- **Briefly describe any working relationship you have (had) with the package authors.**
- [x] As the reviewer I confirm that there are no [conflicts of interest](https://devguide.ropensci.org/policies.html#coi) for me to review this work (if you are unsure whether you are in conflict, please speak to your editor _before_ starting your review).

#### Documentation

The package includes all the following forms of documentation:

- [x] **A statement of need:** clearly stating problems the software is designed to solve and its target audience in README
- [x] **Installation instructions:** for the development version of package and any non-standard dependencies in README
- [x] **Vignette(s):** demonstrating major functionality that runs successfully locally
- [x] **Function Documentation:** for all exported functions
- [x] **Examples:** (that run successfully locally) for all exported functions
- [ ] **Community guidelines:** including contribution guidelines in the README or CONTRIBUTING, and DESCRIPTION with `URL`, `BugReports` and `Maintainer` (which may be autogenerated via `Authors@R`).

#### Functionality

- [ ] **Installation:** Installation succeeds as documented.
- [x] **Functionality:** Any functional claims of the software been confirmed.
- [ ] **Performance:** Any performance claims of the software been confirmed.
- [x] **Automated tests:** Unit tests cover essential functions of the package and a reasonable range of inputs and conditions. All tests pass on the local machine.
- [x] **Packaging guidelines**: The package conforms to the rOpenSci packaging guidelines.

Estimated hours spent reviewing:

- ~ 9

- [ ] Should the author(s) deem it appropriate, I agree to be acknowledged as a package reviewer ("rev" role) in the package DESCRIPTION file.

---

### Review Comments

- The `canaper` package provides functions to infer the evolutionary processes underlying neoendemism and paleoendemism. The software is an R implementation of the Categorical Analysis of Neo- and Paleo-endemism, which is only available in the [Biodiverse](https://github.com/shawnlaffan/biodiverse) software. 

- I enjoyed reviewing the package, but I found some things that might improve the package.
    - One concern is about parallelization: in the current stage of the package, parallelization calls are done outside the functions (via `future::plan(future::multisession)`). I think it would be better to have parameters for parallelization in functions that allow running analysis in parallel (so calls are done inside them).

Parallelization calls as in the current version of the package
```r
library(canaper) # This package :)
library(future) # For parallel computing
library(tictoc) # For timing things
library(patchwork) # For composing mult-part plots
library(tidyverse) # For data-wrangling and plottin

data(acacia)

plan(multisession,workers=ncores)
set.seed(071421)
acacia_rand_res <- cpr_rand_test(
  acacia$comm, acacia$phy, 
  null_model = "curveball",
  n_reps = 100, n_iterations = 100000,
  tbl_out = TRUE)
plan(sequential)
```

I think this could be more appropriate and easy to use

```r
acacia_rand_res <- cpr_rand_test(
  acacia$comm, acacia$phy, 
  null_model = "curveball",
  n_reps = 100, n_iterations = 100000,
  tbl_out = TRUE
  parallel = TRUE, # Logical var to run processes in parallel
  nworkes = 3, # Number of threads
  random_seed=071421) # Random seed
```

- Instead of suggesting the `future` package I would import it

#### README

- No code of conduct and contribution guidelines, please add it

#### Installation

There is a potential conflict in package dependencies as `canaper` depends on R (>= 3.5.0), but it imports `phangorn` version 2.8.1, which depends on R (>= 4.1.0). Please change parameter **Depends: R (>= 3.5.0)**  to **Depends: R (>= 4.1.0)** in the description file or provide a way to install an earlier version of `phangorn` (version  2.5.5 should work).

- I installed the package locally on macOS (Big Sur), Linux (Ubuntu 20.04.3 & 21.04), and Windows 10

#### Automated test

- `devtools::check()` takes more than six minutes. Building of vignette outputs takes a lot of time (5m 5.2s). Consider using lees number of replicates in examples (i.e. **n_reps** of `canaper::cpr_rand_test`).
    - I got one error after upgrading from GEOS 3.9 to 3.10 in Linux. The issue is related to phyloregion dependency.
```r
══ Warnings ════════════════════════════════════════════════════════════════════
── Warning (test-utils.R:143:3): Functions copied from phyloregion work ────────
rgeos: versions of GEOS runtime 3.10.1-CAPI-1.16.0
and GEOS at installation 3.9.1-CAPI-1.14.2differ
Backtrace:
	1. testthat::expect_equal(dense2sparse(biod_example$comm), phyloregion::dense2sparse(biod_example$comm)) test-utils.R:143:2
4. base::loadNamespace(x)
7. base::loadNamespace(j <- i[[1L]], c(lib.loc, .libPaths()), versionCheck = vI[[j]])
8. base:::runHook(".onLoad", env, package.lib, package)
13. rgeos:::fun(libname, pkgname)
```
    

- `devtools::test()` finished with no fails and one warning 

```r
────────────────────────────────────────────────────────────────────
Warning (test-utils.R:143:3): Functions copied from phyloregion work
rgeos: versions of GEOS runtime 3.10.1-CAPI-1.16.0
and GEOS at installation 3.9.1-CAPI-1.14.2differ
Backtrace:
  1. testthat::expect_equal(dense2sparse(biod_example$comm), phyloregion::dense2sparse(biod_example$comm)) test-utils.R:143:2
  4. base::loadNamespace(x)
  7. base::loadNamespace(j <- i[[1L]], c(lib.loc, .libPaths()), versionCheck = vI[[j]])
  8. base:::runHook(".onLoad", env, package.lib, package)
 13. rgeos:::fun(libname, pkgname)
────────────────────────────────────────────────────────────────────

══ Results ═════════════════════════════════════════════════════════
Duration: 71.8 s

[ FAIL 0 | WARN 1 | SKIP 0 | PASS 255 ]
```

- `devtools::spell_check()` Change "signficance" to "significance" in line 132

```r
signficance       canape.Rmd:132
```

#### Performance 

When benchmarking, I found a strange behavior. After using more than four threads, the program takes more time to finish a randomization test. Here is the code I used (could you please check it? I might be missing something)

```r
library(purrr)
library(tidyverse)
library(tictoc)
library(future)
library(ggplot2)
rm(list = ls())
data(acacia)
acom <- acacia$comm
acom[acom>0] <- 1
n_reps_vec <- c(100,500,1000,1500,2000)
nwokers <- 1:7
acom <- acacia$comm
acom[acom>0] <- 1

rand_times <- seq_along(n_reps_vec) %>% purrr::map_dfr(function(nrep){
  rand_t <- nwokers %>% purrr::map_dfr(function(nwork){
      if(nwork>1){
        plan(multisession,workers=nwork)
      }
      set.seed(071421)
      tictoc::tic() 
      # Run randomization test
      acacia_rand_res <- canaper::cpr_rand_test(
        acom, acacia$phy,
        null_model = "curveball",
        n_reps = n_reps_vec[nrep],
        n_iterations = 100000,
        tbl_out = TRUE)
      aa <- tictoc::toc()
      r1 <-data.frame(n_reps = n_reps_vec[nrep] ,
                      nworkers=nwork,
                      time_secs = aa$toc - aa$tic)
      plan(sequential)
      return(r1)
    })
  return(rand_t)
})

ggplot(rand_times, aes(x = ncores,y=time_secs,color=as.factor(n_reps))) +
  geom_line() + theme_classic()
```

#### Questions and comments

- I was wondering if `canaper` has already all methods of the CANAPE version of [Biodiverse](http://shawnlaffan.github.io/biodiverse/). Could write something about it in the README?
- Although the vignettes are very useful and well documented, it would be nice to have methods for plotting spatial data results.

@joelnitta please do not hesitate to contact me if you have any questions. Thanks for developing `canaper`.

**Comment by joelnitta** [comment]
@luismurao thanks so much for your review! I will try to address your comments soon.

**Comment by joelnitta** [comment]
@KlausVigo it would be great to have your comments as well so I can make changes incorporating both sets of reviews.

**Comment by KlausVigo** [comment]
Hi @joelnitta! I will finish it today.

**Comment by joelnitta** [comment]
@KlausVigo thank you very much for the extensive and helpful review!

I just wanted to address one thing first: the reason I copied functions from phyloregion instead of importing them is because of a dependency of phyloregion (betapart), which has a package in `Suggests` (snow) that was resetting the random number generator and thus `cpr_rand_test()` was failing to return identical results when using parallel computing, even when `set.seed()` was used. There is an [issue](https://github.com/joelnitta/canaper/issues/2) describing the problem in canaper, and a related [issue](https://github.com/HenrikBengtsson/future/discussions/554) in the future package.

In the end, I opted to copy the phyloregion functions under the AGPL-3 license, since that fixed the issue and resulted in fewer dependencies. As @HenrikBengtsson [mentioned](https://github.com/HenrikBengtsson/future/discussions/554) though, it seems another way to fix this (and import phyloregion normally) would be to use `future.packages = "snow"` with `future.apply::future_lapply()`. Would you prefer that?

**Comment by tdhock** [comment]
hi @KlausVigo thanks for your detailed review. can you please tell us approximately how many hours you spent on reviewing? (you left that field blank in your review comment)

**Comment by tdhock** [comment]
@ropensci-review-bot submit review https://github.com/ropensci/software-review/issues/475#issuecomment-1002754119 time 9

**Comment by KlausVigo** [comment]
Hi @joelnitta,  

I just made a commit to phyloregion [cb819fb](https://github.com/darunabas/phyloregion/commit/cb819fb430e9c539d36f74937e2c680575d5e0e7). This might solves the problem,  at least `sessionInfo()` suggests that `snow` is not loaded via a namespace any more.

Regards, 
Klaus

> @KlausVigo thank you very much for the extensive and helpful review!
> 
> I just wanted to address one thing first: the reason I copied functions from phyloregion instead of importing them is because of a dependency of phyloregion (betapart), which has a package in `Suggests` (snow) that was resetting the random number generator and thus `cpr_rand_test()` was failing to return identical results when using parallel computing, even when `set.seed()` was used. There is an [issue](https://github.com/joelnitta/canaper/issues/2) describing the problem in canaper, and a related [discussion](https://github.com/HenrikBengtsson/future/discussions/554) in the future package.
> 
> In the end, I opted to copy the phyloregion functions under the AGPL-3 license, since that fixed the issue and resulted in fewer dependencies. As @HenrikBengtsson [mentioned](https://github.com/HenrikBengtsson/future/discussions/554) though, it seems another way to fix this (and import phyloregion normally) would be to use `future.packages = "snow"` with `future.apply::future_lapply()`. Would you prefer that?
> 

> (By the way, it is possible that this behavior of snow may cause reproducibility issues with phyloregion as well. I suggest you look into it if you haven't already).

**Comment by joelnitta** [comment]
@KlausVigo thanks for making that change: [I switched back to importing phyloregion](https://github.com/joelnitta/canaper/commit/7e8b24950912bfa2a5e10cf4dd9cfa8910d74e76), and it [passes the tests](https://github.com/joelnitta/canaper/commit/0c95b885e76abf563b70099a814d62744f6ee65c) for consistent results using the same seed in parallel.

I just have one request: would you consider bumping the phyloregion version so that I can require the version with this fix in `Imports`? That will prevent this bug popping up for others.

> Hi @joelnitta,
> 
> I just made a commit to phyloregion [cb819fb](https://github.com/darunabas/phyloregion/commit/cb819fb430e9c539d36f74937e2c680575d5e0e7). This might solves the problem, at least `sessionInfo()` suggests that `snow` is not loaded via a namespace any more.
>

**Comment by KlausVigo** [comment]
@joelnitta nice that this works and it seems to simplify things. 
I wasn't aware that `ImportFrom(betapart, beta.pair)` and `betaprt::beta.pair` made such a large difference with the NAMESPACE. I naively assumed it is about the same and the documentation about these issues does not tell you otherwise. @darunabas is the maintainer of `phyloregion` and has to submit it to CRAN. We probably need a week or so to clean up all changes and get the error messages sorted out as we haven't updated the package for a while. 


> @KlausVigo thanks for making that change: [I switched back to importing phyloregion](https://github.com/joelnitta/canaper/commit/7e8b24950912bfa2a5e10cf4dd9cfa8910d74e76), and it [passes the tests](https://github.com/joelnitta/canaper/commit/0c95b885e76abf563b70099a814d62744f6ee65c) for consistent results using the same seed in parallel.
> 
> I just have one request: would you consider bumping the phyloregion version so that I can require the version with this fix in `Imports`? That will prevent this bug popping up for others.

**Comment by joelnitta** [comment]
@KlausVigo no problem: it is a very subtle error, and one that I would never have caught without the help of @HenrikBengtsson. As [he has mentioned](https://twitter.com/henrikbengtsson/status/1460452409776357382) this is likely a much wider problem across many packages on CRAN. 

If you and @darunabas are able to release a new version of `phyloregion` in a reasonable amount of time on CRAN that would be great.

**Comment by shawnlaffan** [comment]
@KlausVigo  - if you are updating ```phyloregion``` then it would be good to also address https://github.com/darunabas/phyloregion/issues/3

**Comment by tdhock** [comment]
I am wondering if we can move forward on this submission? @KlausVigo @luismurao have the comments/suggestions in your reviews been addressed?

**Comment by joelnitta** [comment]
@tdhock Sorry, I am the cause for the delay! Since the reviews came back I've gotten busy with other projects, and haven't been able to get back to this yet. I anticipate being occupied by other projects for another week or two. Please remind me again if you don't hear something soon after that.

**Comment by emilyriederer** [comment]
Hi @joelnitta and @tdhock  ! I'm currently the rotating editor in chief for rOpenSci submissions. I wanted to check in and see if you expect to have time to revisit the review feedback soon? Otherwise, we may want to put this review on hold status

**Comment by joelnitta** [comment]
Thanks for checking in @emilyriederer. Sorry for falling behind on this. I am still occupied with other things at the moment, but I think I should be able to respond to the reviews by the end of August if not sooner.

**Comment by joelnitta** [comment]
## Response to reviews

I am sorry for taking so long to respond to the reviewers. Thanks again to both reviewers for their helpful comments. Below I respond to each comment as appropriate.

### Response to @luismurao

- [x] Concern about parallelization. I think it would be better to have parameters for parallelization in functions that allow running analysis in parallel (so calls are done inside them).
- **Response:** The second reviewer, @KlausVigo disagreed with this point and was in favor of leaving the current approach to parallelization, so I am making no changes.

- [x] Instead of suggesting the `future` package I would import it
- **Response:** I disagree with this. Parallelization is not required, it is an option to speed up analyses if desired. So I think it makes sense as `Suggests` instead of `Imports`. No changes made.

#### README

- [x] No code of conduct and contribution guidelines, please add it
- **Response:** There was apparently a misunderstanding; contribution guidelines are already provided at `/.github/CONTRIBUTING.md`. To make it easier to find though, I have added a link to this document from the README. I also added a link to the ROpenSci code of conduct using the [suggested wording in the developer guide](https://devguide.ropensci.org/collaboration.html?q=code%20of%20c#coc-file) ([d8aaf8de5193166d52ab890d8267090ea2f8ef6a](https://github.com/joelnitta/canaper/commit/d8aaf8de5193166d52ab890d8267090ea2f8ef6a)).

#### Installation

- [x] Please change parameter Depends: R (>= 3.5.0) to Depends: R (>= 4.1.0) in the description file or provide a way to install an earlier version of phangorn (version 2.5.5 should work).
- **Response:** I changed to `Depends: R (>= 4.1.0)` ([0f334965fb5a3fe5034eb1bc682b5051548eddb1](https://github.com/joelnitta/canaper/commit/0f334965fb5a3fe5034eb1bc682b5051548eddb1)).

#### Automated test

- [x] `devtools::check()` takes more than six minutes. Building of vignette outputs takes a lot of time (5m 5.2s). Consider using lees number of replicates in examples (i.e. `n_reps` of `canaper::cpr_rand_test`).
- **Response:** I had to include a non-trivial number replicates to produce reasonable results (e.g., in comparison with Mishler et al. 2014 for the [CANAPE example vignette](https://joelnitta.github.io/canaper/articles/canape.html)) and to demonstrate performance gains from parallelization. But I was able to decrease the number of replicates in the vignettes ([f3af4671d679020aeaa97b1a038b3b5a96633f42](https://github.com/joelnitta/canaper/commit/f3af4671d679020aeaa97b1a038b3b5a96633f42)). The vignettes now take 3m 40s on my computer to compile.

- [x] I got one error after upgrading from GEOS 3.9 to 3.10 in Linux. The issue is related to phyloregion dependency.
- **Response:** Yes: phylogregion imports `rgeos`, which is [going to be retired soon](https://r-spatial.org/r/2022/04/12/evolution.html) and [needs fixing](https://github.com/darunabas/phyloregion/issues/11). Correct installation of external dependencies to phyloregion is outside the scope of `canaper`. No change made (for now). 

- [x] `devtools::test()` finished with no fails and one warning
- **Response:** On my computer, `devtools::test()` now finishes with no fails or warnings (at [93543d399fd34dd6546e56368986a8afa133f190](https://github.com/joelnitta/canaper/commit/93543d399fd34dd6546e56368986a8afa133f190)).

- [x] `devtools::spell_check()` Change "signficance" to "significance" in line 132
- **Response:** Fixed ([b04e2e0a8397568bcc615e4aaeb5a0d9b02ad85e](https://github.com/joelnitta/canaper/commit/b04e2e0a8397568bcc615e4aaeb5a0d9b02ad85e))

#### Performance

- [x] When benchmarking, I found a strange behavior. After using more than four threads, the program takes more time to finish a randomization test.
- **Response:** Speed is not expected to increase linearly with the number of cores. It is well documented in other software that increasing cores above a certain point causes code to run more slowly due to the resources required to coordinate threads. It is up to the user to identify the optimal number of threads to use. I provide a vignette article discussing precisely this topic: https://joelnitta.github.io/canaper/articles/parallel.html.

#### Questions and comments

- [x] I was wondering if canaper has already all methods of the CANAPE version of Biodiverse. Could write something about it in the README?
- **Response:** I have added a section in the README comparing `canaper` other software ([d8aaf8de5193166d52ab890d8267090ea2f8ef6a](https://github.com/joelnitta/canaper/commit/d8aaf8de5193166d52ab890d8267090ea2f8ef6a)).

- [x] Although the vignettes are very useful and well documented, it would be nice to have methods for plotting spatial data results.
- **Response:** Thanks for the suggestion. I have already included [color palettes](https://joelnitta.github.io/canaper/reference/index.html#color-palettes) for plotting results that should be accessible regardless of color vision type. IMHO there are too many customizations that users would want to make to try and implement a one-size-fits-all plotting method. Rather, the results are structured in a way that they should be easily plotted with e.g. ggplot2, and I [demonstrate this in the vignettes](https://joelnitta.github.io/canaper/articles/canape.html#visualize-results).

### Response to @KlausVigo

- [x] There are some functions in the `picante` (`ses.pd`) and in the `PhyloMeasures` package for Faith's phylogenetic diversity with a similar scope.
- **Response:** There are many R packages that calculate community diversity metrics, and many of them overlap in common metrics like Faith's PD. In the case of `canaper`, I need to calculate Faith's PD as part of the CANAPE calculation. I use the `phyloregion::PD()` to do this as it is the fastest AFAIK. No changes needed.

- [x] The canaper package builds on functions from the vegan package to resample (dense) community matrices and borrows biodiversity measures (Faith's phylogenetic diversity and phylogenetic endemism) from the phyloregion package. The picante package also has some functions for swapping (community) matrices similar to vegan. A comparison between the packages would be useful.
- **Response:** I have added a section to the README comparing `canaper` with other software ([d8aaf8de5193166d52ab890d8267090ea2f8ef6a](https://github.com/joelnitta/canaper/commit/d8aaf8de5193166d52ab890d8267090ea2f8ef6a)).

#### Specific comments

- [x] The functions from the phyloregion package used in the package should be acknowledged properly... An alternative is importing the functions from the package directly.
- **Response:** Please see response [above](https://github.com/ropensci/software-review/issues/475#issuecomment-1047206275). Functions from the phyloregion package are now imported, not copied ([7e8b24950912bfa2a5e10cf4dd9cfa8910d74e76](https://github.com/joelnitta/canaper/commit/7e8b24950912bfa2a5e10cf4dd9cfa8910d74e76)). Please note that phyloregion is still at [v1.0.7 as of writing](https://github.com/darunabas/phyloregion/tree/9b47a325054e9a1b801598a0469e472eefa6be53) (same version as before @KlausVigo [implemented the change](https://github.com/ropensci/software-review/issues/475#issuecomment-1060658515) that fixed the loading of `snow`), so I can't specify the version of phyloregion that would avoid canaper [issue #2](https://github.com/joelnitta/canaper/issues/2).

- [x] In contrast to the other reviewer I really like the use of the `future` package.
- **Response:** I agree, and have left parallelization as it is.

#### Profiling

- [x] Providing a swapping algorithm which can handle sparse matrices would be a huge improvement for the community.
- **Response:** I agree implementing sparse matrix encoding for generating random matrices would be a huge benefit and will consider it as a feature to be added possibly in the future (any help would be appreciated!).

- [x] The functions `calc_biodiv_random` and `cpr_rand_comm` are only called internally and are called thousands of times inside of the loop. It should be save to remove all (most) of the calls to assertthat::assert_that from these functions as parameters have been checked already and are save to use!
- **Response:** Actually, `cpr_rand_comm()` is exported, but point well-taken. I have deleted the redundant checks ([bccd0278dd3de5853680d9979ca26c1aa38a7fb3](https://github.com/joelnitta/canaper/commit/bccd0278dd3de5853680d9979ca26c1aa38a7fb3)).

#### For submitting to CRAN

- [x] There are some instances where more than 2 cores are used. This could fail on CRAN, when submitting.
- **Response:** Thanks for pointing this out. Fixed ([12bbe36bf5516a2a63c1235c76234708c4c5432d](https://github.com/joelnitta/canaper/commit/12bbe36bf5516a2a63c1235c76234708c4c5432d), [f3af4671d679020aeaa97b1a038b3b5a96633f42](https://github.com/joelnitta/canaper/commit/f3af4671d679020aeaa97b1a038b3b5a96633f42))

### General comments

- Both reviewers left `Performance: Any performance claims of the software been confirmed` unchecked. I am not sure why this is the case, as I include tests of increase in performance using parallel computing **and** demonstrate this feature in a [vignette](https://joelnitta.github.io/canaper/articles/parallel.html).

- I have added @KlausVigo as a reviewer, but not @luismurao, according to their responses to `Should the author(s) deem it appropriate, I agree to be acknowledged as a package reviewer`.

**Comment by joelnitta** [comment]
## Update to response to reviewers

The longest running test is [`"Parallelization decreases calculation time"`](https://github.com/joelnitta/canaper/blob/76c277490dcecd0f32df351c48c8f8c891674aad/tests/testthat/test-cpr_rand_test.R#L391). To make this test robust, I have to include a relatively high number of replicates, otherwise by random chance sometimes running in parallel won't be faster than running sequentially.

I have decided to follow [6.1.5.4 Extended tests](https://stats-devguide.ropensci.org/standards.html#extended-tests) of [rOpenSci Statistical Software Peer Review](https://stats-devguide.ropensci.org/standards.html) and treat this as an extended test that can be optionally skipped (it is now skipped by default).

I have [documented this in a comment](https://github.com/joelnitta/canaper/blob/76c277490dcecd0f32df351c48c8f8c891674aad/tests/testthat/test-cpr_rand_test.R#L395) as follows:

```
  # This test is skipped by default because of its long run time.
  # To include the test locally, first run
  # withr::local_envvar(CANAPER_EXTENDED_TESTS = "true") # nolint
  # To include the test in CI (github actions), include the phrase
  # 'run-extended' in the commit
```

`devtools::test()` on my computer now finishes in 36 s skipping the extended test, and 181 s when including it.

@tdhock, @emilyriederer sorry again for the long delay, but this should be ready to move forward. Please let me know if there is anything you need from my end.

**Comment by emilyriederer** [comment]
Also tagging  @KlausVigo @luismurao . I know this review was a bit delayed and your availability may have changed. At your convenience, if you could please review @joelnitta 's responses and share whether these address your feedback ([template available here](https://devguide.ropensci.org/reviewerguide.html#followupreviewer)) , that would be fantastic.

**Comment by joelnitta** [comment]
@emilyriederer If the reviewers have no further comments, can we proceed with this at your discretion as an editor?

**Comment by emilyriederer** [comment]
Hi @joelnitta - let me follow up with @tdhock since he's the managing editor on this one

**Comment by tdhock** [comment]
hi! Since the reviewers have been tagged two weeks ago and gave no feedback, I think it is OK to move ahead with this submission (the reviewer response is very complete/convincing).

**Comment by tdhock** [comment]
@ropensci-review-bot approve canaper

**Comment by joelnitta** [comment]
Thanks @tdhock!

**Comment by joelnitta** [comment]
@ropensci-review-bot finalize transfer of canaper
